### ISSUE TITLE
Potential improvement to reverse_range_loop

### ISSUE THREAD

**Issue by pzmarzly** [rust-lang-nursery/rust-clippy#2477]
While it is stated that this lint cannot catch loops over dynamically defined ranges, it could catch the following common pattern:

```rust
let x = 5u32;
for i in x..0 { ... }
```

Note that it is a range over unsigned values, so it makes no sense for the right side to be zero (as the range will always be empty then). Similar check could be done for signed integers by `iXXX::MIN` const.

**Comment by mxxo** [comment]
I don't want to hijack this issue, but clippy also misses adapters on invalid ranges like: 
```rust 
// skips loop 
for i in (10..0).rev() {
         println!("{}", i);
}
```

**Comment by Zalathar** [comment]
I ran into a similar issue with a backwards loop that modifies a vector:

```rust
// Wrong!
for i in (vec.len()-1)..=0 {
    // Code that structurally modifies vec,
    // so we need backwards traversal
    // and can’t use a slice iterator.
}
```

The correct range is `(0..vec.len()).rev()`, which also conveniently avoids unsigned overflow when `vec` is empty.

**Comment by giraffate** [comment]
`reverse_range_loop` is now included in `reversed_empty_ranges`, so I changed this title.
